### ISSUE TITLE
Restructuring CDash Testing

### ISSUE THREAD

**Issue by danlamanna** [Kitware/CDash#280]
Leaving an issue here to capture some thoughts I've had/bounced off of @zackgalbreath.

CDash maintains shared state across its tests, this has already caused me a couple of headaches and is a bit counter intuitive. It breaks the local development environment configs after running, and the tests are dependent on the order in which they're run - hindering the ability to parallelize them.

Contrast this to an approach similar to Girder's, where CherryPy is setup/tore down and a database is created for each individual test.  If we were to implement something similar it would give us isolated and more easily parallelized testing environments.

A few things that might help with this, I'd like some input on:
- Using the built in PHP dev server for testing?  
  I've been using this for dev and it's worked well and although I'm sure there are shortcomings, it might be useful for testing. 
- Letting env variables specify configurations  
  If configurations/configuration files could be specified as an environment variable then it would be easier to spin up multiple instances of CDash from the same codebase utilizing different databases or whatever instance specific configuration we want.

**Comment by jamiesnape** [comment]
For the filesystem: `composer require mikey179/vfsStream`

**Comment by jamiesnape** [comment]
While we are thinking of this, we should use `phpunit/phpunit` and `mockery/mockery`. The current framework is flaky and outdated.

**Comment by zackgalbreath** [comment]
We discussed this a bit more today.  Here's what we decided to try.  @danlamanna / @bryonbean, let me know if I failed to capture anything important.

Our goal is to be able to run each test independently.  If test 100 is failing, you shouldn't have to run 1-99 first.  We also want to be able to run our tests in parallel.

* Each test will be responsible for creating a new database and submitting any necessary data to it.
* Each test that needs to interact with a webserver (all of them?) will start up and tear down PHP's built-in server (using exec).  We will have to assign different port numbers per test somehow (manually?)
* We will call `genev` at the bottom of `config.php` and override the values of any `$CDASH_` variables found.  This should give us a more sane way of overriding configuration at runtime.
* We will begin this effort towards the end of our list of tests.  If that goes well, we will gradually march our way back towards the beginning.
* We will audit what coverage is gained by the `simple` tests (and any other than invoke a c/c++ compiler).  These seem out of scope for what CDash is meant to do.

**Comment by zackgalbreath** [comment]
> Each test that needs to interact with a webserver (all of them?) will start up and tear down PHP's built-in server (using exec).

I've recently spent some time towards this goal (an effort started by @danlamanna).  At this point I've got a good number of the tests passing under this system.

A problem I've run into is that the built-in server is single-threaded.  Some of our tests exercise explicitly parallel interactions, particularly `test_processsubmissions` and `test_parallelsubmissions`.  I'm not sure what to do about these.

`test_actualtrilinossubmission` passes, but it takes a lot longer to finish: 253s vs. 117s for master.

This is a bit of a bummer.  My plan was to lay the initial ground work.  Then we could subsequently update tests in batches to run in isolation.  But leaving our testing environment in a worse state (even temporarily) is not ideal.  This is more of a status update than a query, but I welcome any thoughts or suggestions you might have upon reading this news.

**Comment by zackgalbreath** [comment]
As we go through this on a test-by-test basis, we should take a closer look that test_recoverpassword is actually doing what we think it should (#530)

**Comment by williamjallen** [comment]
All recent tests have been deliberately designed to be independent, and I've made several recent changes to the test suite to make legacy tests run independently.  Given that we now have a solution for this problem, I'll close this issue in favor of more specific issues.
